app/src/foundation/ContentScript/ContentScript.ts

```diff
@@ -21,64 +21,71 @@ interface Current {
+    message: any,
+    devtoolsPort: Runtime.Port,
+  ) {
     if (message?.type === 'accept-port' && message?.sender === 'dev') {
-      return setup(port);
+      return setup({devtoolsPort});
     }
   }
 
-  async function onConnectListener(port: Runtime.Port) {
-    port.postMessage({type: 'accept-port', sender: 'content-script'});
+  async function onConnectListener(devtoolsPort: Runtime.Port) {
+    devtoolsPort.postMessage({type: 'accept-port', sender: 'content-script'});
 
     const listener = (message: any) => {
       if (message.type === 'ready' && message.sender === 'dev') {
-        setup(port);
+        setup({devtoolsPort});
       }
     };
-    port.onMessage.addListener(listener);
+    devtoolsPort.onMessage.addListener(listener);
   }
 
-  async function setup(port: Runtime.Port) {
-    // Unmount the client side remote-components.
-    // This wipes the devtools UI completely.
+  async function setup({devtoolsPort}: {devtoolsPort?: Runtime.Port}) {
     try {
       await current.webpage?.call.unmountDevtools();
     } catch (error) {
       console.log('webpage terminate failed');
     }
 
-    // Create a brand new devtools endpoint
-    const devtools = createDevtoolsEndpoint(port);
-
-    // Use the same webpage endpoint or create a new one if there's none.
-    const webpage = current.webpage ?? createWebpageEndpoint();
-
-    // Even if the webpage already existed, the methods exposed to the webpage
-    // depend on the new dev tools endpoint, so we need to re-expose all methods anyway.
-    exposeWebpage(webpage, devtools);
-    exposeDevtools(devtools, webpage);
-
-    port.onDisconnect.addListener(() => {
-      if (current.port === port) {
-        current.port = undefined;
-      }
-      current.webpage?.call.unmountDevtools();
-    });
-
-    current.devtools = devtools;
-    current.webpage = webpage;
-    current.port = port;
-
-    webpage.call.mountDevtools();
+    if (devtoolsPort) {
+      // Create a brand new devtools endpoint
+      const devtools = createDevtoolsEndpoint(devtoolsPort);
+
+      // Use the same webpage endpoint or create a new one if there's none.
+      const webpage = current.webpage ?? createWebpageEndpoint();
+
+      // Even if the webpage already existed, the methods exposed to the webpage
+      // depend on the new dev tools endpoint, so we need to re-expose all methods anyway.
+      exposeWebpage(webpage, devtools);
+      exposeDevtools(devtools, webpage);
+
+      devtoolsPort.onDisconnect.addListener(() => {
+        if (current.port === devtoolsPort) {
+          current.port = undefined;
+        }
+        current.webpage?.call.unmountDevtools();
+      });
+
+      current.devtools = devtools;
+      current.webpage = webpage;
+      current.port = devtoolsPort;
+
+      webpage.call.mountDevtools();
+    } else {
+      const webpage = current.webpage ?? createWebpageEndpoint();
+      exposeWebpage(webpage);
+      current.webpage = webpage;
```

Comment:
Change here is to allow the setup to be started without having a devtools opened. This allows the content script to start communicating with the webpage and the background/service-worker.

Reply:
Previously, the content script was receiving a request to connect from the main page but we were not creating the content-script -> webpage endpoint until the devtools was actually opened.